stateresolutionv2.go

```diff
@@ -0,0 +1,483 @@
+			if mainlineEvent.EventID() == searchEvent.EventID() {
+				return true, pos
+			}
+		}
+		// If we've reached this point then the event is not in the mainline.
+		return false, 0
+	}
+
+	// Define our iterator function.
+	var iter func(event Event)
+	iter = func(event Event) {
+		// In much the same way as we do in createPowerLevelMainline, we loop
+		// through the event's auth events, checking that it exists in our supplied
+		// auth event map and finding power level events.
+		for _, authEventID := range event.AuthEventIDs() {
+			// Check that we actually have the auth event in our map - we need this so
+			// that we can look up the event type.
+			if authEvent, ok := r.authEventMap[authEventID]; ok {
+				// Is the event a power level event?
+				if authEvent.Type() == MRoomPowerLevels {
+					// Is the event in the mainline?
+					if isIn, pos := isInMainline(authEvent); isIn {
+						// It is - take a note of the event and position and stop the
+						// iterator from running any further.
+						mainlineEvent = authEvent
+						mainlinePosition = pos
+						return
+					}
+					// It isn't - increase the step count and then run the iterator again
+					// from the found auth event.
+					steps++
+					iter(authEvent)
+				}
+			}
+		}
+	}
+
+	// Start the iterator with the supplied event.
+	iter(event)
+
+	return
+}
+
+// authAndApplyEvents iterates through the supplied list of events and auths
+// them against the current partial state. If they pass the auth checks then we
+// also apply them on top of the partial state.
+func (r *stateResolverV2) authAndApplyEvents(events []Event) {
+	for _, e := range events {
+		event := e
+		// Check if the event is allowed based on the current partial state. If the
+		// event isn't allowed then simply ignore it and process the next one.
+		if err := Allowed(event, r); err != nil {
+			continue
+		}
+		// We've now authed the event - work out what the type is and apply it to
+		// the partial state based on type.
+		switch event.Type() {
+		case MRoomCreate:
+			// Room creation events are only valid with an empty state key.
+			if event.StateKey() == nil || *event.StateKey() == "" {
+				r.resolvedCreate = &event
+			}
+		case MRoomPowerLevels:
+			// Power level events are only valid with an empty state key.
+			if event.StateKey() == nil || *event.StateKey() == "" {
+				r.resolvedPowerLevels = &event
+			}
+		case MRoomJoinRules:
+			// Join rule events are only valid with an empty state key.
+			if event.StateKey() == nil || *event.StateKey() == "" {
+				r.resolvedJoinRules = &event
+			}
+		case MRoomThirdPartyInvite:
+			// Third party invite events are only valid with a non-empty state key.
+			if event.StateKey() != nil && *event.StateKey() != "" {
+				r.resolvedThirdPartyInvites[*event.StateKey()] = &event
+			}
+		case MRoomMember:
+			// Membership events are only valid with a non-empty state key.
+			if event.StateKey() != nil && *event.StateKey() != "" {
```

Comment:
It would be a definite violation of spec for these checks to fail and the room version 2 doc states:
> If the state event is not allowed by the authorization rules, then the event is ignored. 

We don't return an error at this point because otherwise the state resolution algorithm would be left in a partial state.

Reply:
SGTM, the spec seems clear on this.